README.md
# base16-builder-typescript

A Base16 builder for node written in Typescript

## Installation

`npm install -g base16-builder-typescript`

## Requirements

git

## Usage

Supports [base16 builder spec 1.0.0](https://github.com/chriskempson/base16/blob/main/builder.md).

Example usage: `cat tomorrow-night.yaml | base16-builder --template vim.mustache > tomorrow-night.vim`

### Other Usage

First pull down all known schemes and templates  
`base16-builder update`  
This will pull down known sources into `sources` and then will clone schemes and templates into `sources/schemes` and `sources/templates` respectively. If schemes and templates are already cloned then they will be updated. 

Then build the themes  
`base16-builder build`  
Themes will be placed in `themes` in the directory where you ran `base16-builder`.

## Additional Features

`base16-builder build --template [template]`  
This will build with only the specified template. The template names are specified [here](https://github.com/chriskempson/base16-templates-source/blob/master/list.yaml). For example `base16-builder build --template textmate` will build only the Textmate templates with all schemes.

`base16-builder build --scheme [scheme]`  
This will build with only the specified scheme. The scheme names are specified [here](https://github.com/base16-project/base16-schemes). For example `base16-builder build --scheme default-dark` will build all templates with only the default dark scheme.

These options can be used with each other. For example `base16-builder build --template textmate --scheme default-dark` will only build the Textmate theme with only the default schemes.

## References

[Builder Guidelines](https://github.com/chriskempson/base16/blob/main/builder.md)  
[Base16 Repository](https://github.com/chriskempson/base16)  


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2022-08-01
### Changed
- Changed schemes source to [base16-project/base16-schemes](https://github.com/base16-project/base16-schemes) to bring back Tomorrow themes. Doing this caused a breaking change in the `--scheme` parameter. Instead of passing in the scheme folder you now must pass in the name of the scheme file.

### Fixed
- Fixed bug where options wouldn't be used.

## [1.2.1] - 2022-08-01
### Fixed
- Fixed bug where crash would happen if template config.yaml didn't exist.

## [1.2.0] - 2022-07-29
### Added
- Support for [base16 builder spec 1.0.0](https://github.com/chriskempson/base16/blob/main/builder.md)
  - Can now pipe in a scheme file and specify a template file to output the application theme to stdout.

## [1.1.3] - 2017-08-18
### Fixed
- Change line endings to LF to fix running as a binary on Unix systems.

## [1.1.2] - 2017-07-29
### Fixed
- Prepend `base16` to rendered filename.

## [1.1.1] - 2017-07-29
### Changed
- Use filename for `scheme-slug` instead of `scheme-name`. 0.9.0 builder requirement.

## [1.1.0] - 2017-07-16
### Added
- Option to build with a specified template or specified scheme

## [1.0.0] - 2017-07-16
### Added
- Initial release


src/update.ts
import * as fs from 'fs';
import * as path from 'path';
import * as jsyaml from 'js-yaml';
import * as mkdirp from 'mkdirp';
import * as shelljs from 'shelljs';

export function update() {
    let sourcesDir = path.resolve('.', 'sources');
    if (!fs.existsSync(sourcesDir)) {
        try {
            mkdirp.sync(sourcesDir);
        } catch (err) {
            console.error(err);
            process.exit(1);
        }
    }

    let sources: any = jsyaml.load(fs.readFileSync(path.resolve(__dirname, '../../sources.yaml'), 'utf8'));
    gitUpdate(sources, sourcesDir);

    // Only gitUpdate schemes if the repository is from chriskempson.
    // let schemesDir = path.resolve(sourcesDir, 'schemes');
    // let schemes: any = jsyaml.load(fs.readFileSync(path.resolve(schemesDir, 'list.yaml'), 'utf8'));
    // gitUpdate(schemes, schemesDir);

    let templatesDir = path.resolve(sourcesDir, 'templates');
    let templates: any = jsyaml.load(fs.readFileSync(path.resolve(templatesDir, 'list.yaml'), 'utf8'));
    gitUpdate(templates, templatesDir);
}

function gitUpdate(yaml: any, dir: string): void {
    if (!yaml) {
        return;
    }
    let keys = Object.keys(yaml);
    keys.forEach(k => {
        let keyDir = path.resolve(dir, k);
        if (!fs.existsSync(keyDir)) {
            try {
                mkdirp.sync(keyDir);
            } catch(err) {
                console.error(err);
                process.exit(1);
            }
            shelljs.exec(`git clone ${yaml[k]} ${keyDir}`);
        }
        else {
            shelljs.exec(`git pull ${keyDir}`);
        }
    });
}


src/builder.ts
import * as fs from 'fs';
import * as path from 'path';
import * as jsyaml from 'js-yaml';
import * as mkdirp from 'mkdirp';
import * as shelljs from 'shelljs';
import * as mustache from 'mustache';
import * as color from 'color';
const slugify = require('slugify');

export function builder(options?: any) {
    let sourcesDir = path.resolve('.', 'sources');
    if (!fs.existsSync(sourcesDir)) {
        console.error('Could not find sources directory. Please run base16-builder update first. Exiting.');
        process.exit(1);
    }

    let schemesDir = path.resolve(sourcesDir, 'schemes');
    let schemes = fs.readdirSync(schemesDir).filter(f => {
        // get scheme file names
        return f.endsWith('.yaml');
    })
    .map(f => {
        // and then remove the extension to make it easier to match with the scheme option
        return f.slice(0, -5);
    });
    let templatesDir = path.resolve(sourcesDir, 'templates');
    let templateFolders = fs.readdirSync(templatesDir).filter(d => {
        return d.indexOf('.') == -1;
    });

    if (options.template && templateFolders.indexOf(options.template) == -1) {
        console.error(`Template ${options.template} does not exist. Exiting.`);
        process.exit(1);
    }
    if (options.scheme && schemes.indexOf(options.scheme) == -1) {
        console.error(`Scheme ${options.scheme} does not exist. Exiting.`);
        process.exit(1);
    }

    let themesDir = path.resolve('.', 'themes');
    if (!fs.existsSync(themesDir)) {
        try {
            mkdirp.sync(themesDir);
        } catch (err) {
            console.error(err);
            process.exit(1);
        }
    }
    shelljs.rm('-rf', `${themesDir}/*`);

    // for each template folder (alacritty, c_header, crosh, etc.)
    templateFolders.forEach(templateFolder => {
        if (options.template && templateFolder != options.template) {
            return;
        }
        let currentTemplateDirectory = path.resolve(templatesDir, templateFolder);
        // load the config in template_folder/templates/config.yaml
        let configFile: string;
        try {
            configFile = fs.readFileSync(path.resolve(currentTemplateDirectory, 'templates/config.yaml'), 'utf8');
        } catch (err) {
            // Config file does not exist.
            return;
        }
        let config: any = jsyaml.load(configFile);
        // get the keys inside the template folder (default, default-256, etc.)
        let configKeys = Object.keys(config);
        // then for each key
        configKeys.forEach(key => {
            // create the output directory at themes/template_folder/output
            let outputDir = path.join(themesDir, templateFolder, config[key].output);
            if (!fs.existsSync(outputDir)) {
                try {
                    mkdirp.sync(outputDir);
                } catch (err) {
                    console.error(err);
                    process.exit(1);
                }
            }
            // grab the file extension
            let fileExtension = config[key].extension;
            // then for each scheme folder
            schemes.forEach(scheme => {
                if (options.scheme && scheme != options.scheme) {
                    return;
                }
                let currentSchemeFile = path.resolve(schemesDir, `${scheme}.yaml`);
                // load the scheme file
                let yaml: any = jsyaml.load(fs.readFileSync(currentSchemeFile, 'utf8'));
                if (!yaml.base00) {
                    // test checking if scheme file is properly formatted
                    return;
                }
                // create the view
                let view = createView(yaml);
                // load the mustache file
                let mustacheFile = fs.readFileSync(path.resolve(currentTemplateDirectory, 'templates', `${key}.mustache`), 'utf8');
                // render the file
                let renderedFile = mustache.render(mustacheFile, view);
                // and write it out to the output directory with the appropriate name
                try {
                    const outputFile = path.resolve(outputDir, `base16-${view['scheme-slug']}${fileExtension}`);
                    fs.writeFileSync(outputFile, renderedFile);
                } catch (err) {
                    console.error(err);
                }
            });
        });
    });
}

export function buildFromPipe(scheme: string, options?: any) {
    if (!options.template) {
        console.error('Please specify a template file.');
        process.exit(1);
    }
    const yaml: any = jsyaml.load(scheme);
    if (!yaml.base00) {
        console.error('Scheme file is not properly formatted.');
        process.exit(1);
    }
    const mustacheFile = fs.readFileSync(options.template, 'utf8');
    const renderedFile = buildTemplate(yaml, mustacheFile);
    process.stdout.write(renderedFile);
}

function buildTemplate(schemeYaml: any, mustacheFile: string): string {
    let view = createView(schemeYaml);
    let renderedFile = mustache.render(mustacheFile, view);
    return renderedFile;
}

function createView(yaml: any): object {
    let final = {};
    final['scheme-name'] = yaml.scheme;
    final['scheme-author'] = yaml.author;
    final['scheme-slug'] = slugify(yaml.scheme).toLowerCase();
    const baseIds = [
        'base00',
        'base01',
        'base02',
        'base03',
        'base04',
        'base05',
        'base06',
        'base07',
        'base08',
        'base09',
        'base0A',
        'base0B',
        'base0C',
        'base0D',
        'base0E',
        'base0F'
    ];
    baseIds.forEach(id => {
        final[`${id}-hex`] = yaml[id];
        final[`${id}-hex-bgr`] = yaml[id].slice(4, 6) + yaml[id].slice(2, 4) + yaml[id].slice(0, 2);
        final[`${id}-hex-r`] = yaml[id].slice(0, 2);
        final[`${id}-hex-g`] = yaml[id].slice(2, 4);
        final[`${id}-hex-b`] = yaml[id].slice(4, 6);
        final[`${id}-rgb-r`] = color(`#${yaml[id]}`).red();
        final[`${id}-rgb-g`] = color(`#${yaml[id]}`).green();
        final[`${id}-rgb-b`] = color(`#${yaml[id]}`).blue();
        final[`${id}-dec-r`] = color(`#${yaml[id]}`).red() / 255;
        final[`${id}-dec-g`] = color(`#${yaml[id]}`).green() / 255;
        final[`${id}-dec-b`] = color(`#${yaml[id]}`).blue() / 255;
        final[`${id}-hsl-h`] = color(`#${yaml[id]}`).hsl().hue();
        final[`${id}-hsl-s`] = color(`#${yaml[id]}`).hsl().saturationl();
        final[`${id}-hsl-l`] = color(`#${yaml[id]}`).hsl().lightness();
    });
    return final;
}


out/src/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = void 0;
const fs = require("fs");
const path = require("path");
const jsyaml = require("js-yaml");
const mkdirp = require("mkdirp");
const shelljs = require("shelljs");
function update() {
    let sourcesDir = path.resolve('.', 'sources');
    if (!fs.existsSync(sourcesDir)) {
        try {
            mkdirp.sync(sourcesDir);
        }
        catch (err) {
            console.error(err);
            process.exit(1);
        }
    }
    let sources = jsyaml.load(fs.readFileSync(path.resolve(__dirname, '../../sources.yaml'), 'utf8'));
    gitUpdate(sources, sourcesDir);
    // Only gitUpdate schemes if the repository is from chriskempson.
    // let schemesDir = path.resolve(sourcesDir, 'schemes');
    // let schemes: any = jsyaml.load(fs.readFileSync(path.resolve(schemesDir, 'list.yaml'), 'utf8'));
    // gitUpdate(schemes, schemesDir);
    let templatesDir = path.resolve(sourcesDir, 'templates');
    let templates = jsyaml.load(fs.readFileSync(path.resolve(templatesDir, 'list.yaml'), 'utf8'));
    gitUpdate(templates, templatesDir);
}
exports.update = update;
function gitUpdate(yaml, dir) {
    if (!yaml) {
        return;
    }
    let keys = Object.keys(yaml);
    keys.forEach(k => {
        let keyDir = path.resolve(dir, k);
        if (!fs.existsSync(keyDir)) {
            try {
                mkdirp.sync(keyDir);
            }
            catch (err) {
                console.error(err);
                process.exit(1);
            }
            shelljs.exec(`git clone ${yaml[k]} ${keyDir}`);
        }
        else {
            shelljs.exec(`git pull ${keyDir}`);
        }
    });
}
//# sourceMappingURL=update.js.map

src/base16-builder.ts
import * as commander from 'commander';
import * as shelljs from 'shelljs';
import * as update from './update';
import * as builder from './builder';

if (!shelljs.which('git')) {
    console.error('Could not find git on path. Exiting.');
    process.exit(1);
}

const chunks = [];

const program = new commander.Command();
program.version('1.0.0');
program.command('build')
    .description('builds all themes')
    .option('-t, --template [template]', 'build with only the specified template')
    .option('-s, --scheme [scheme]', 'build with only the specified scheme')
    .action(function() {
        const options = program.opts();
        builder.builder(options);
    });
program.command('update')
    .description('clones or pulls sources, schemes, and template repositories')
    .action(() => {
        update.update();
    });
program
    .option('-t, --template [template file]')
    .action(function() {
        const options = program.opts();
        if (process.stdin.isTTY) {
            if (process.argv.length === 2) {
                program.help();
            } else {
                console.error('Please pipe in a scheme file.');
                process.exit(1);
            }
        } else {
            process.stdin.on('readable', () => {
                let chunk;
                while ((chunk = process.stdin.read()) !== null) {
                    chunks.push(chunk);
                }
            });
        
            process.stdin.on('end', () => {
                const content = chunks.join('');
                builder.buildFromPipe(content, options);
            });
        }
    });

program.parse(process.argv);


out/src/base16-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commander = require("commander");
const shelljs = require("shelljs");
const update = require("./update");
const builder = require("./builder");
if (!shelljs.which('git')) {
    console.error('Could not find git on path. Exiting.');
    process.exit(1);
}
const chunks = [];
const program = new commander.Command();
program.version('1.0.0');
program.command('build')
    .description('builds all themes')
    .option('-t, --template [template]', 'build with only the specified template')
    .option('-s, --scheme [scheme]', 'build with only the specified scheme')
    .action(function () {
    const options = program.opts();
    builder.builder(options);
});
program.command('update')
    .description('clones or pulls sources, schemes, and template repositories')
    .action(() => {
    update.update();
});
program
    .option('-t, --template [template file]')
    .action(function () {
    const options = program.opts();
    if (process.stdin.isTTY) {
        if (process.argv.length === 2) {
            program.help();
        }
        else {
            console.error('Please pipe in a scheme file.');
            process.exit(1);
        }
    }
    else {
        process.stdin.on('readable', () => {
            let chunk;
            while ((chunk = process.stdin.read()) !== null) {
                chunks.push(chunk);
            }
        });
        process.stdin.on('end', () => {
            const content = chunks.join('');
            builder.buildFromPipe(content, options);
        });
    }
});
program.parse(process.argv);
//# sourceMappingURL=base16-builder.js.map

out/src/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildFromPipe = exports.builder = void 0;
const fs = require("fs");
const path = require("path");
const jsyaml = require("js-yaml");
const mkdirp = require("mkdirp");
const shelljs = require("shelljs");
const mustache = require("mustache");
const color = require("color");
const slugify = require('slugify');
function builder(options) {
    let sourcesDir = path.resolve('.', 'sources');
    if (!fs.existsSync(sourcesDir)) {
        console.error('Could not find sources directory. Please run base16-builder update first. Exiting.');
        process.exit(1);
    }
    let schemesDir = path.resolve(sourcesDir, 'schemes');
    let schemes = fs.readdirSync(schemesDir).filter(f => {
        // get scheme file names
        return f.endsWith('.yaml');
    })
        .map(f => {
        // and then remove the extension to make it easier to match with the scheme option
        return f.slice(0, -5);
    });
    let templatesDir = path.resolve(sourcesDir, 'templates');
    let templateFolders = fs.readdirSync(templatesDir).filter(d => {
        return d.indexOf('.') == -1;
    });
    if (options.template && templateFolders.indexOf(options.template) == -1) {
        console.error(`Template ${options.template} does not exist. Exiting.`);
        process.exit(1);
    }
    if (options.scheme && schemes.indexOf(options.scheme) == -1) {
        console.error(`Scheme ${options.scheme} does not exist. Exiting.`);
        process.exit(1);
    }
    let themesDir = path.resolve('.', 'themes');
    if (!fs.existsSync(themesDir)) {
        try {
            mkdirp.sync(themesDir);
        }
        catch (err) {
            console.error(err);
            process.exit(1);
        }
    }
    shelljs.rm('-rf', `${themesDir}/*`);
    // for each template folder (alacritty, c_header, crosh, etc.)
    templateFolders.forEach(templateFolder => {
        if (options.template && templateFolder != options.template) {
            return;
        }
        let currentTemplateDirectory = path.resolve(templatesDir, templateFolder);
        // load the config in template_folder/templates/config.yaml
        let configFile;
        try {
            configFile = fs.readFileSync(path.resolve(currentTemplateDirectory, 'templates/config.yaml'), 'utf8');
        }
        catch (err) {
            // Config file does not exist.
            return;
        }
        let config = jsyaml.load(configFile);
        // get the keys inside the template folder (default, default-256, etc.)
        let configKeys = Object.keys(config);
        // then for each key
        configKeys.forEach(key => {
            // create the output directory at themes/template_folder/output
            let outputDir = path.join(themesDir, templateFolder, config[key].output);
            if (!fs.existsSync(outputDir)) {
                try {
                    mkdirp.sync(outputDir);
                }
                catch (err) {
                    console.error(err);
                    process.exit(1);
                }
            }
            // grab the file extension
            let fileExtension = config[key].extension;
            // then for each scheme folder
            schemes.forEach(scheme => {
                if (options.scheme && scheme != options.scheme) {
                    return;
                }
                let currentSchemeFile = path.resolve(schemesDir, `${scheme}.yaml`);
                // load the scheme file
                let yaml = jsyaml.load(fs.readFileSync(currentSchemeFile, 'utf8'));
                if (!yaml.base00) {
                    // test checking if scheme file is properly formatted
                    return;
                }
                // create the view
                let view = createView(yaml);
                // load the mustache file
                let mustacheFile = fs.readFileSync(path.resolve(currentTemplateDirectory, 'templates', `${key}.mustache`), 'utf8');
                // render the file
                let renderedFile = mustache.render(mustacheFile, view);
                // and write it out to the output directory with the appropriate name
                try {
                    const outputFile = path.resolve(outputDir, `base16-${view['scheme-slug']}${fileExtension}`);
                    fs.writeFileSync(outputFile, renderedFile);
                }
                catch (err) {
                    console.error(err);
                }
            });
        });
    });
}
exports.builder = builder;
function buildFromPipe(scheme, options) {
    if (!options.template) {
        console.error('Please specify a template file.');
        process.exit(1);
    }
    const yaml = jsyaml.load(scheme);
    if (!yaml.base00) {
        console.error('Scheme file is not properly formatted.');
        process.exit(1);
    }
    const mustacheFile = fs.readFileSync(options.template, 'utf8');
    const renderedFile = buildTemplate(yaml, mustacheFile);
    process.stdout.write(renderedFile);
}
exports.buildFromPipe = buildFromPipe;
function buildTemplate(schemeYaml, mustacheFile) {
    let view = createView(schemeYaml);
    let renderedFile = mustache.render(mustacheFile, view);
    return renderedFile;
}
function createView(yaml) {
    let final = {};
    final['scheme-name'] = yaml.scheme;
    final['scheme-author'] = yaml.author;
    final['scheme-slug'] = slugify(yaml.scheme).toLowerCase();
    const baseIds = [
        'base00',
        'base01',
        'base02',
        'base03',
        'base04',
        'base05',
        'base06',
        'base07',
        'base08',
        'base09',
        'base0A',
        'base0B',
        'base0C',
        'base0D',
        'base0E',
        'base0F'
    ];
    baseIds.forEach(id => {
        final[`${id}-hex`] = yaml[id];
        final[`${id}-hex-bgr`] = yaml[id].slice(4, 6) + yaml[id].slice(2, 4) + yaml[id].slice(0, 2);
        final[`${id}-hex-r`] = yaml[id].slice(0, 2);
        final[`${id}-hex-g`] = yaml[id].slice(2, 4);
        final[`${id}-hex-b`] = yaml[id].slice(4, 6);
        final[`${id}-rgb-r`] = color(`#${yaml[id]}`).red();
        final[`${id}-rgb-g`] = color(`#${yaml[id]}`).green();
        final[`${id}-rgb-b`] = color(`#${yaml[id]}`).blue();
        final[`${id}-dec-r`] = color(`#${yaml[id]}`).red() / 255;
        final[`${id}-dec-g`] = color(`#${yaml[id]}`).green() / 255;
        final[`${id}-dec-b`] = color(`#${yaml[id]}`).blue() / 255;
        final[`${id}-hsl-h`] = color(`#${yaml[id]}`).hsl().hue();
        final[`${id}-hsl-s`] = color(`#${yaml[id]}`).hsl().saturationl();
        final[`${id}-hsl-l`] = color(`#${yaml[id]}`).hsl().lightness();
    });
    return final;
}
//# sourceMappingURL=builder.js.map